std/array.d

```diff
@@ -1980,38 +1980,62 @@ if (isInputRange!RoR &&
 /++
     Replace occurrences of `from` with `to` in `subject` in a new
     array. If `sink` is defined, then output the new array into
-    `sink`.
+    `sink`. If `countChanges` is defined, place the number of
+    replacements done in it.
 
     Params:
         sink = an $(REF_ALTTEXT output range, isOutputRange, std,range,primitives)
         subject = the array to scan
         from = the item to replace
         to = the item to replace all instances of `from` with
+        countChanges = the number of substitutions performed
 
     Returns:
         If `sink` isn't defined, a new array without changing the
         contents of `subject`, or the original array if no match
-        is found.
+        is found. The number of replacements is placed in countChanges
+        if it is defined.
 
     See_Also:
         $(REF map, std,algorithm,iteration) which can act as a lazy replace
  +/
-E[] replace(E, R1, R2)(E[] subject, R1 from, R2 to)
-if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
-        && (hasLength!R2 || isSomeString!R2))
+auto replace(E, R1, R2)(E subject, R1 from, R2 to)
+if (isForwardRange!E && isForwardRange!R1 && isForwardRange!R2)
 {
-    import std.algorithm.searching : find;
+    size_t countChanges = 0;
+    return replace(subject, from, to, countChanges);
+}
 
-    if (from.empty) return subject;
+/// ditto
+auto replace(E, R1, R2)(E subject, R1 from, R2 to, out size_t countChanges)
+if (isForwardRange!E && isForwardRange!R1 && isForwardRange!R2)
+{
+    import std.algorithm.searching : findSplit;
 
-    auto balance = find(subject, from.save);
-    if (balance.empty)
+    countChanges = 0;
+    if (from.empty)
         return subject;
 
-    auto app = appender!(E[])();
-    app.put(subject[0 .. subject.length - balance.length]);
-    app.put(to.save);
-    replaceInto(app, balance[from.length .. $], from, to);
+    static if (is(ElementType!E : R1))
+    {
+        import std.range : only;
+        auto res = findSplit(subject, from.only);
+    }
+    else
+    {
+        auto res = findSplit(subject, from);
+    }
+
+    if (res[1].empty)
+        return subject;
+
+    auto app = appender!((ElementEncodingType!E)[])();
+    put(app, res[0]);
+    put(app, to.save);
```

Comment:
Btw @RazvanN7 all the `.save` calls that you have inserted are only necessary if by-reference Ranges are supported. Currently there are no tests for this, hence maybe adding a test for it wouldn't be a bad idea ;-)
Tip: [`refRange`](https://dlang.org/phobos/std_range.html#.refRange) is your friend ;-)